### ISSUE TITLE
NET Core runtime for desktop -> NET Core Desktop Runtime

### ISSUE THREAD

**Issue by ace2k** [Tyrrrz/LightBulb#134]
### Problem
Since the release of 2.0 I've read on several occasions how people presumably follow the instructions yet fail to launch the app succesfully or end up with _(for non-devs)_ cryptic error logs. Upon taking a closer look, there seems to be a misunderstanding in the actual software dependency, which at least partly can be credited to Microsoft's constant change of the naming convention in their software packages.

Most people unfortunately skim over guides/readmes rather than reading patiently, _(which seems to be a side-effect of today's informational overload and the subsequent decrease of attention span)_, only to end up with the wrong installation of dependency packages and blame the software or even the developer.

![NET Runtime Packages Installed](https://s.put.re/8gv6MoSp.png)

---

### Explanation
In order to run LightBulb 2.0, the intstructions explicitly suggest to install "**NET Core runtime for desktop apps**", which is semantically correct but can be easily confused with the baseline package due to the capitalization rules of the words in the name, ie. easily mistaking it for "**NET Core runtime**", which according to Microsoft is for running console applications only.

Whereas the [wiki article](https://github.com/Tyrrrz/LightBulb/wiki/How-to-install-.NET-Core-runtime) directly links to the "**NET Core Desktop Runtime**" download page, which according to Microsoft is the one required to run desktop apps (or console apps).

Ideally this wouldn't be a problem at all if everyone just followed the wiki article's link. For various reasons like the lack of software understanding, lack of attention, language barrier, naming confusion et cetera this is not the case and often results in a bad user experience and subsequent dismissal of the software. I think this can be resolved with rather simple tweaks.

Source: https://dotnet.microsoft.com/download/dotnet-core/current
![NET Runtime Packages](https://s.put.re/4bsfMt8f.png)

---

### Suggestion
Looking forward to run that one app everyone is talking about, only to fail right from the get-go can be ultminately frustrating - even more so if you are not a techie person. For the sake of clarity and in order to make this easier for everyone incl. absolute PC newbies or simply people who skim over guides _(which seem to be a wide-spread phenomenon)_, I suggest to change every text occurence of  **`NET Core runtime`** and **`NET Core runtime for desktop apps`** to **`NET Core Desktop Runtime`**, as suggested by the chosen name on Microsoft's own download page.

Thanks for reading!
_Edit: a word._

**Comment by Tyrrrz** [comment]
Hi, thank you for such a lengthy write up. I tried to make it as clear as possible but I see how this can be confusing to some people.

Also, when LightBulb v2.0 just released, you were required to download both the base runtime and the desktop runtime. Luckily they repackaged it in a single installer later.

I've already thought about it before and recently started again considering doing self-contained deployments. The downside is that the file size will increase from 8.2mb to 100mb, which is a pretty big deal for users with slower internet.

I'm also thinking if I can pack the runtime with the installer (previously it wasn't possible/easy) but then the portable users will be out of luck. On top of that, when LightBulb moves to the next version of the framework, it will screw up auto-update.

Related to #122

**Comment by ace2k** [comment]
Hey, thanks for the clarification about prior requirement of both packages. I dimly remember reading about that but since I already had both runtime packages installed (iirc. v3.0.1) at the time when LightBulb 1.6.4 was still my daily driver, it never really occured to me being an issue until people started reporting problems and asking for help with the launch of LightBulb v2.0.

Hopefully Microsoft can sort this out someday without constantly making unnecessary and confusing changes. Ideally they would provide the runtime as an initial release (v2, v3, v4 etc.) mandatory via Windows Update and offer small subsequent updates in a similar fashion hotfixes are delivered, that can be packaged as a minimal requirement with software installers such as LightBulb, rather than having to clog up the pipe with a 10x larger dependency every single release, which is pretty radical regardless of data caps or internet connection.

I think the setups (incl. portable) in their current compact form are perfect. But that's of course just one point of view and the answer depends on how many people will ultimately run into this issue described above.

**Comment by xcasxcursex** [comment]
TL;DR at the bottom.

I came across this, this week. I recommended Lightbulb as a superior alternative to f.lux (because it is) and the person I spoke to said that he tried it and it didn't work. I immediately recognised that he hadn't installed the .NET runtimes and told him as much. He said something to the effect of "I just ran the installer and started the app and nothing happens, every other app I use works like this so it must be a problem with lightbulb". I mean let's face it: in no less than three places, at the top of the pages, right next to the download links, in bold letters, highlighted paragraphs, step-by-step wiki instructions, with links to them elsewhere, people are informed of the need to install the runtimes.... It should be enough. I thought it was more than enough....but clearly, I've overestimated the average person's ability. Sorry if that seems harsh but I'm just sayin'.... People aren't that smart.

Perhaps I should have been more delicate and polite about the cause of this problem and said it like OP " (which seems to be a side-effect of today's informational overload and the subsequent decrease of attention span)" but after double-checking the process and seeing no less than three reminders of the need for the runtimes right next to the downloads I guess I'm a little less compassionate about it all ;)


But still, it seems crazy to have to make a 100MB installer for a 10MB app. I have really nice fast internet so it doesn't bother me at all but I can absolutely see why you've opted for the lighter installer. Not only is the smaller installer sensible for users with slower internet (Thinking about my friends in outback Australia, not to mention COVID-19 grinding all our internet to a halt) it also makes a lot more sense for portable installs - if I've installed the runtimes in the OS, then I don't need them packaged in the app in a portable install as well. But then again, I guess there are two flavours of 'portable' - one where the app itself is portable, assuming the OS is fully operational and includes the latest runtimes....and one where the app is portable regardless of the state of the OS, so the portable app is also carrying part of the OS as 'portable'.


I remember early in the transition to 2.0, you had a system where the installer for lightbulb would launch a download of the runtime installer. This seems to be a good way to go. It allows for a small distribution of the app, and then if the installer can detect if the runtime exists, and it doesn't, then download and launch the runtime installer.

Alternatively, if the installer itself cannot detect the presence of the runtimes, perhaps the app itself could do this. That way, if people are too dense to read one line of text right above and below the download link, it will compensate for them by saying like "Hey, you're missing this windows component, click this button to download it now".

That just leaves the last use-case though, where people want to run a portable install which does not require the runtimes to be installed in the OS at all and wants them included in the app.


**TL;DR It seems like you need a few different options for installers:**

Small installer which (either in the installer or the app itself) automatically detects the runtime presence and downloads and runs the installer if necessary.
Small portable version which behaves the same way.
Big portable installer which includes the runtimes.

PS Murphy's Law says that as soon as you get this working, Microsoft will begin to package the runtimes as part of the OS (like they should) and the entire problem will go away, LOL.

**Comment by Tyrrrz** [comment]
Hey @xcasxcursex thanks for your comment. I really appreciate your detailed insights.

I'm not surprised there are people who simply don't pay attention or don't care enough to take extra steps to install software, especially if they weren't very motivated to do it in the first place (e.g. when someone recommends it to them but they didn't particularly want to install it themselves). I'm not sure what to make of it or quite frankly if it's a bad thing at all.

To be clear, I want to make the installation process as smooth as possible for all users, of course. Previously, with .NET Framework there were no additional installs because it was packaged with Windows, but I had to purposely target the lowest version I could so that it overlapped with the pre-installed versions on oldest versions of Windows. It was quite limiting. Also, if Microsoft ever decides to package .NET runtimes with Windows, I'm 99% sure it will only concern latest versions of Windows 10, so some users who are still on Windows 7 and use LightBulb because there is no built-in alternative, are going to have to install the runtime manually anyway.

In other news, I updated the installer today so that it downloads and installs .NET runtime automatically, if it's not already installed. It's kind of awkward because there are these 3 distinctly separate stages: LightBulb installer, PowerShell script, .NET Runtime installer, but at least the user just has to follow along and they will be fine.

I tried including this script with the original 2.0 release, but back then the distribution story was really horrible, you actually needed two separate installers instead of one. I just couldn't get it to work in time before my self-imposed release deadline so I scrapped it.

I think the future is with self-contained deployments. Currently it's just not really there yet. This September, Microsoft is releasing .NET 5 which, among other things, will improve self-contained deployments so that they take up less space. If, say, the changes would let me bring down LightBulb distro to like 30mb, I would be okay with that. We'll see.

Currently, there's also another elephant in the room - autoupdate. My biggest issue right now is that I can't target a higher framework version because the autoupdate doesn't actually use the installer, it simply updates the files. That means it won't be able to update the runtime. This is also why I'll probably make it so the next release targets .NET Core 3.0 (like the last one) instead of 3.1 (as it is in master), just to avoid disruption. Hopefully the self-contained deployment story is much better in .NET 5 and then all of these issues are magically fixed.

**Comment by Tyrrrz** [comment]
I've also considered adding extra download options. The problem is that it also requires the users will read and download what they need. Which, if they did, they would've been fine either way. :)

**Comment by xcasxcursex** [comment]
Like the saying goes: You can't patch stupid.

I think you're doing the best you can with the problems at hand.....

Anyway I tested the new CI build after uninstalling the framework and it downloaded a fresh one and works fine.

**Comment by xcasxcursex** [comment]
Just to really throw a spanner in the works, I uninstalled the .net runtimes AFTER using the new build which had automatically installed them.

When I ran lightbulb, it still prompted me to go download the runtimes, which I did, and all is working as expected. So that's nice :)



Possibly of note: The automatic install from the powershell script in the latest Lightbulb CI build, installed runtime version 3.1.2.
When I got the prompt from the app and followed it to[ the download page](https://dotnet.microsoft.com/download/dotnet-core/?utm_source=getdotnetcore&utm_medium=referral), the latest and recommended and marked as LTS version, was 3.1.3

Not gonna lie, once you get to that page, there are about 5 bazillion different links and the correct one is not immediately apparent. Even knowing full well which one I needed, I managed to click the wrong link (because it was the first one in the big box and I was not paying attention like a dumbie) and install the SDK. which works with Lighbulb of course, but is overkill ;) I then almost accidentally downloaded the ASP.NET Core Runtime before I realised I was about to screw up again and eventually got to the Desktop Runtime. Not to be confused with the .NET Core Runtime. WHAT A MESS, MICROSOFT! (Take a look at that download page, you'll see what I mean)


Annnyway, the upside of this is that after clicking the correct link for the correct runtime, it was kind enough to provide me a direct link that I thought I might share with you:
x64:
https://download.visualstudio.microsoft.com/download/pr/5954c748-86a1-4823-9e7d-d35f6039317a/169e82cbf6fdeb678c5558c5d0a83834/windowsdesktop-runtime-3.1.3-win-x64.exe
Checksum: 
9e0c3ab51de6265d8ade1235f9cd6d5a9932459cfff6327fe09b9e9342f0b54c4d41bfc7bcc29c1e3e8994c5df7cfba530126a6ee06de3c32ea3cfcc36013a5f

x86:
https://download.visualstudio.microsoft.com/download/pr/7cd5c874-5d11-4e72-81f0-4a005d956708/0eb310169770c893407169fc3abaac4f/windowsdesktop-runtime-3.1.3-win-x86.exe
Checksum:
f56d72723864ad55b5a238d07a6b34ac3bcefe92c646aefb8b09cd00db54450192b88982cfc057eee7a6039fc41733da2ab9d626e6dde0240e009a494060699e

Since these are supposedly LTS versions I figure they might be useful for the PowerShell script? Or not... I just thought I'd save you some time getting those links if you need them.

**Comment by Tyrrrz** [comment]
v3.1.3 was released right after I created the script :)
The patch version doesn't really matter as long as it's 3.1, but if you want you can make a PR to update the URLs in the script.

>By the way, you mentioned rolling back to 3.0 which is now marked EOL. I swear, Microsoft are deliberately trying to make this difficult :P

That's true. The app will target .NET Core 3.0 but I will still point everyone to install 3.1. That way it will work just fine. The only reason I'm considering doing this is because there are quite a lot of people who downloaded LightBulb in the first week or two and only have .NET Core 3.0 -- if I make a new release, the auto-update will overwrite the files and will make the app unusable until they download a new version of the runtime.

**Comment by xcasxcursex** [comment]
>  if you want you can make a PR to update the URLs in the script.

I want to help, so I tried.....but at work we don't use git, so I am a complete noob. I hope I didn't totally screw it up XD

> the auto-update will overwrite the files and will make the app unusable until they download a new version of the runtime.

I assume that microsoft in their endless attempts to make this difficult, have not provided a way to detect the runtime version and update that, too.
I think you're doing well in the face of MS's very poorly implemented runtime update.

**Comment by ace2k** [comment]
Apparently **.NET Framework 4.8** and **.Net Core 3** are being consolidated into **.NET 5.0** in November 2020.
